feature: adding forget password feature

basharwalid opened the pull request:
adding forget password feature to the app which include three main screens:
.forget password screen 
. OTP verify upon the user associated account email
. reset password screen for the user to set a new password

basharwalid committed:
+ forget password feature:
 - forget password email
 - otp send

basharwalid committed:
- verify reset code sent by api
- ui to be implemented

basharwalid committed:
- verify reset code sent by api
- ui to be implemented

basharwalid committed:
+ otp check done
 - resend button to be implemented
 - reset password screen to be implemented

basharwalid committed:
feat: add forget password feature in the app which this commit includes:
. forget password screen
. otp verify screen
. reset password screen

basharwalid committed:
Merge remote-tracking branch 'origin/development' into development

# Conflicts:
#	lib/core/l10n/app_ar.arb
#	lib/core/l10n/app_en.arb
#	lib/core/theme/app_theme.dart
#	lib/data/datasource/impl/auth_remote_datasource_impl.dart
#	lib/data/repository/auth_repository_impl.dart
#	lib/domain/repository/auth_repository.dart
#	lib/main.dart
#	lib/presentation/login/login_contract.dart
#	lib/presentation/login/widget/login_form.dart
#	lib/presentation/signup/signup_view.dart
#	lib/presentation/signup/widgets/signup_form.dart

basharwalid committed:
(fix): removed locale duplicates

john-safwat commented on lib/data/models/authentication/verify_reset_code/request/verify_reset_code_request_dto.dart:
```diff
@@ -0,0 +1,30 @@
+      VerifyResetCodeRequestDto(resetCode: request.resetCode);
+}
+
+
```
unnecessary model for jus string

john-safwat commented on lib/data/datasource/impl/auth_remote_datasource_impl.dart:
```diff
@@ -40,4 +49,40 @@ class AuthRemoteDatasourceImpl implements AuthRemoteDatasource {
+
+  @override
+  Future<Results<VerifyResetCodeResponse>> verifyResetCode(
+      VerifyResetCodeRequest verify) async {
```
replace model with just string no need to create model for one string variable

john-safwat commented on lib/data/models/authentication/forget_password/request/forget_password_request_dto.dart:
```diff
@@ -0,0 +1,25 @@
+
+  factory ForgetPasswordRequestDto.fromDomain(ForgetPasswordRequest request) =>
+      ForgetPasswordRequestDto(email: request.email);
+}
```
no need for this model just pass the string parameter

john-safwat commented on lib/data/datasource/impl/auth_remote_datasource_impl.dart:
```diff
@@ -40,4 +49,40 @@ class AuthRemoteDatasourceImpl implements AuthRemoteDatasource {
+  @override
+  Future<Results<ResetPasswordResponse>> resetPassword(
+      ResetPasswordRequest request) async {
+    var response = await _apiExecution.execute<ResetPasswordResponse>(() async {
+      var response = await _authRetrofitClient.resetPassword(
+          ResetPasswordRequestDto(
+              email: request.email, newPassword: request.newPassword));
+      return response.toDomain();
+    });
+    return response;
```
replace with factorty constractor to convert from entity to model

john-safwat commented on lib/data/models/authentication/reset_password/request/reset_password_request_dto.dart:
```diff
@@ -0,0 +1,26 @@
+    return _$ResetPasswordRequestDtoToJson(this);
+  }
+}
+
```
add factory constractor to convert from entoty to model

john-safwat commented on lib/data/models/authentication/verify_reset_code/request/verify_reset_code_request_dto.dart:
```diff
@@ -0,0 +1,30 @@
+    return _$VerifyResetCodeRequestDtoToJson(this);
+  }
+
+  // factory ForgetPasswordRequestDto.fromDomain(ForgetPasswordRequest request) =>
+  //     ForgetPasswordRequestDto(email: request.email);
```
don''t leave commented code

john-safwat commented on lib/data/models/authentication/verify_reset_code/response/verify_reset_code_response_dto.dart:
```diff
@@ -0,0 +1,28 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:online_exam_app/domain/entities/authentication/verify_reset_code/verify_reset_code_response.dart';
+
+part 'verify_reset_code_response_dto.g.dart';
+
+@JsonSerializable()
+class VerifyResetCodeResponseDto {
+  @JsonKey(name: "status")
+  final String? status;
+
+  VerifyResetCodeResponseDto ({
+    this.status,
+  });
+
+  factory VerifyResetCodeResponseDto.fromJson(Map<String, dynamic> json) {
+    return _$VerifyResetCodeResponseDtoFromJson(json);
+  }
+
+  Map<String, dynamic> toJson() {
+    return _$VerifyResetCodeResponseDtoToJson(this);
+  }
+
+  VerifyResetCodeResponse toDomain() => VerifyResetCodeResponse(
+     status: status,
+  );
+}
+
+
```
no need for model for just string

john-safwat commented on lib/domain/entities/authentication/forgetPassword/forget_password_request.dart:
```diff
@@ -0,0 +1,7 @@
+class ForgetPasswordRequest {
+  final String? email;
+
+  ForgetPasswordRequest({
+    this.email,
+  });
+}
```
no need for model for just string

john-safwat commented on lib/domain/entities/authentication/verify_reset_code/verify_reset_code_request.dart:
```diff
@@ -0,0 +1,7 @@
+class VerifyResetCodeRequest {
+  final String? resetCode;
+
+  VerifyResetCodeRequest({
+    this.resetCode,
+  });
```
unnecessary model

john-safwat commented on lib/domain/entities/authentication/verify_reset_code/verify_reset_code_response.dart:
```diff
@@ -0,0 +1,7 @@
+class VerifyResetCodeResponse {
+  final String? status;
+
+  VerifyResetCodeResponse({
+    this.status,
+  });
+}
```
unnecessary model

john-safwat commented on lib/domain/use_case/forget_password_use_case.dart:
```diff
@@ -0,0 +1,18 @@
+  final AuthRepository _repository;
+
+  const ForgetPasswordUseCase(this._repository);
+
+  Future<Results<ForgetPasswordResponse>> call(
+      ForgetPasswordRequest request) async {
+    var response = await _repository.forgetPassword(request);
+    return response;
+  }
+}
```
you can replace with 
Future<Results<ForgetPasswordResponse>> call(
      ForgetPasswordRequest request) async => await _repository.forgetPassword(request);

john-safwat commented on lib/presentation/forget_password/Widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,63 @@
+        child: ListView(
+          padding: const EdgeInsets.all(16),
+          children: [
+            Row(
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: [
+                Text(viewModel.locale!.forgetPassword ,
+                  style: AppTheme.appTheme.textTheme.titleLarge!.copyWith(fontSize: 18)
+                ),
+              ],
+            ),
```
you can replace the row with center

john-safwat commented on lib/presentation/forget_password/Widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,63 @@
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: [
+                Text(viewModel.locale!.pleaseEnterYourEmailAssociatedToYourAccount ,
+                    style: AppTheme.appTheme.textTheme.titleLarge!.copyWith(fontSize: 14)
```
replace apptheme with Theme.of(context)

john-safwat commented on lib/presentation/forget_password/forget_password_view.dart:
```diff
@@ -0,0 +1,74 @@
+                message: viewModel.locale!.otpSentToYourEmail,
+                context: context,
+                posActionTitle: viewModel.locale!.ok,
+                posAction: () {
+                  Navigator.pushReplacementNamed(context, Routes.otpVerifyRoute,
+                    arguments: viewModel.emailController.text
+                  );
+                });
+          }
```
this is mvvm violation because this code make view apply logic you should make action then call this action for navigation and pass the arguments in the state emitted

john-safwat commented on lib/presentation/forget_password/forget_password_view_model.dart:
```diff
@@ -0,0 +1,83 @@
+  void _forgetPassword() async {
+    if (formKey.currentState!.validate()) {
+      emit(ForgetPasswordLoadingState());
+      var response = await _forgetPasswordUseCase
+          .call(ForgetPasswordRequest(email: emailController.text));
+      emit(HideLoadingState());
```
replace this code with this 
var response = await _forgetPasswordUseCase(ForgetPasswordRequest(email: emailController.text));

john-safwat commented on lib/presentation/otp_verify/Widget/otp_form.dart:
```diff
@@ -0,0 +1,86 @@
+              style:
+                  AppTheme.appTheme.textTheme.bodyLarge!.copyWith(fontSize: 14),
+            ),
+          ],
+        ),
+        const SizedBox(
+          height: 16,
+        ),
+        Pinput(
+          controller: viewModel.otpController,
+          length: 6,
+          defaultPinTheme: defaultPinTheme,
+          focusedPinTheme: defaultPinTheme.copyWith(
+            decoration: defaultPinTheme.decoration!
+                .copyWith(border: Border.all(color: AppColors.green)),
+          ),
+          onCompleted: (value) =>
+              viewModel.doIntent(OnCompleteCodeVerifyAction()),
+        ),
+        Row(
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [
+            Text(
+              viewModel.locale!.didNotReceiveCode,
+              style:
+                  AppTheme.appTheme.textTheme.bodyLarge!.copyWith(fontSize: 14),
+            ),
+            TextButton(
+                onPressed: () {},
+                child: Text(
+                  viewModel.locale!.resend,
+                  style: AppTheme.appTheme.textTheme.bodyLarge!
+                      .copyWith(fontSize: 14),
+                ))
+          ],
+        ),
+      ],
+    );
+  }
+}
```
don't use otp package

john-safwat commented on lib/presentation/otp_verify/otp_verify_view.dart:
```diff
@@ -0,0 +1,66 @@
+class _OtpVerifyViewState extends BaseState<OtpVerifyView, OtpVerifyViewModel> {
+  @override
+  Widget build(BuildContext context) {
+    final email = ModalRoute.of(context)?.settings.arguments as String?;
```
move email to viewModel

john-safwat commented on lib/presentation/otp_verify/otp_verify_view.dart:
```diff
@@ -0,0 +1,66 @@
+                context: context,
+                posActionTitle: viewModel.locale!.ok,
+                posAction: () {
+                  Navigator.pushReplacementNamed(
+                      context, Routes.resetPasswordViewRoute , arguments: email);
+                });
```
MVVM Violation

john-safwat commented on lib/presentation/otp_verify/otp_verify_view_model.dart:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+
+  void _resendOtp() async {
+    // emit(OtpVerifyLoadingState());
+    // var response =
+    //     await _forgetPasswordUseCase.call(ForgetPasswordRequest(email: ""));
+    // emit(HideLoadingState());
+    //
+    // switch (response) {
+    //   case Success<ForgetPasswordResponse>():
+    //     {}
+    //   case Failure<ForgetPasswordResponse>():
+    //     {}
+    // }
+  }
+}
```
remote commented code

john-safwat reviewed (commented):
ovaer all great jop

basharwalid committed:
(fix):
- OTP form UI changed to Text field
- Models that depends on single variable changed
- removed OTP verify request
